### PR TITLE
Add clang format & skip build if no source file modified

### PR DESCRIPTION
**Pull request by nerzhul**
Warning: it check the whole modified file, not the diff part, it's why it's lazy. Please also look if rules are perfect, i take the Linux codestyle from LLVM site

Fix issue #5415 
ping @sfan5 

I also skip build if we didn't modified any CPP file as there is nothing to test

### CODE REVIEW

File: util/travis/script.sh [hunk 1]
```diff
@@ -3,21 +3,61 @@
...
+if [ $(git diff --name-only --diff-filter=ACMRTUXB $TRAVIS_COMMIT_RANGE | grep -c '^src/[^.]*[.]\(c\|cpp\|h\)$') -eq 0 ]; then
+	exit 0
+fi
+
 if [[ $PLATFORM == "Unix" ]]; then
 	mkdir -p travisbuild
 	cd travisbuild || exit 1
+	if [[ "$LINT" == "1" ]]; then
+		# Lint with exit CI
+        perform_lint
```
Comment (sfan5) [inline util/travis/script.sh]: indent

Comment (nerzhul) [inline util/travis/script.sh]: oops, exact

File: util/travis/script.sh [hunk 2]
```diff
@@ -3,21 +3,61 @@
...
+function perform_lint() {
+	CLANG_FORMAT=clang-format-3.9
+	${CLANG_FORMAT} --version
+
+	if [ "$TRAVIS_EVENT_TYPE" = "pull_request" ]; then
+		# Get list of every file modified in this pull request
+		files_to_lint="$(git diff --name-only --diff-filter=ACMRTUXB $TRAVIS_COMMIT_RANGE | grep '^src/[^.]*[.]\(cpp\|h\)$' || true)"
+	else
+		# Check everything for branch pushes
+		files_to_lint="$(find src/ -name '*.cpp' -or -name '*.h')"
```
Comment (sfan5) [inline util/travis/script.sh]: this should skip `gmp`, `jsoncpp` and `lua`

Comment (nerzhul) [inline util/travis/script.sh]: fixed

File: util/travis/script.sh [hunk 3]
```diff
@@ -3,21 +3,61 @@
...
+function perform_lint() {
+	CLANG_FORMAT=clang-format-3.9
+	${CLANG_FORMAT} --version
+
+	if [ "$TRAVIS_EVENT_TYPE" = "pull_request" ]; then
+		# Get list of every file modified in this pull request
+		files_to_lint="$(git diff --name-only --diff-filter=ACMRTUXB $TRAVIS_COMMIT_RANGE | grep '^src/[^.]*[.]\(cpp\|h\)$' || true)"
+	else
+		# Check everything for branch pushes
+		files_to_lint="$(find src/ -name '*.cpp' -or -name '*.h' | egrep -v "src/(gmp|lua|jsoncpp)")"
```
Comment (sfan5) [inline util/travis/script.sh]: minor thing, this needs to be `"^src/(gmp|lua|jsoncpp)"`

File: util/travis/before_install.sh [hunk 4]
```diff
@@ -21,6 +21,12 @@ if [[ $PLATFORM == "Unix" ]]; then
 		if [[ "$VALGRIND" == "1" ]]; then
 			sudo apt-get install valgrind
 		fi
+		if [[ "$LINT" == "1" ]]; then
+			curl http://apt.llvm.org/llvm-snapshot.gpg.key | sudo -E apt-key add -
+			sudo -E add-apt-repository "deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-3.9 main"
```
Comment (sfan5) [inline util/travis/before_install.sh]: the `-E` is not needed is it?

File: util/travis/script.sh [hunk 5]
```diff
@@ -3,21 +3,61 @@
 
 needs_compile || exit 0
 
+function perform_lint() {
+	CLANG_FORMAT=clang-format-3.9
+	${CLANG_FORMAT} --version
+
+	if [ "$TRAVIS_EVENT_TYPE" = "pull_request" ]; then
+		# Get list of every file modified in this pull request
+		files_to_lint="$(git diff --name-only --diff-filter=ACMRTUXB $TRAVIS_COMMIT_RANGE | grep '^src/[^.]*[.]\(cpp\|h\)$' || true)"
```
Comment (sfan5) [inline util/travis/script.sh]: doesn't that need to be `egrep` here?

Comment (nerzhul) [inline util/travis/script.sh]: yes, i will add it

File: util/travis/script.sh [hunk 6]
```diff
@@ -3,21 +3,61 @@
...
+	else
+		# Check everything for branch pushes
+		files_to_lint="$(find src/ -name '*.cpp' -or -name '*.h' | egrep -v "src/(gmp|lua|jsoncpp)")"
+	fi
+
+	for f in ${files_to_lint}; do
+		d=$(diff -u "$f" <(${CLANG_FORMAT} "$f") || true)
+		if ! [ -z "$d" ]; then
+			echo "$f not compliant to coding style, here is the fix:"
+			echo "$d"
```
Comment (sfan5) [inline util/travis/script.sh]: I suggest replacing with:
`printf "The file %s is no compliant with the coding style:\n%s\n" "$f" "$d"`

File: util/travis/script.sh [hunk 7]
```diff
@@ -3,21 +3,61 @@
...
+
+if [ $(git diff --name-only --diff-filter=ACMRTUXB $TRAVIS_COMMIT_RANGE | grep -c '^src/[^.]*[.]\(c\|cpp\|h\)$') -eq 0 ]; then
+	exit 0
+fi
+
 if [[ $PLATFORM == "Unix" ]]; then
 	mkdir -p travisbuild
 	cd travisbuild || exit 1
+	if [[ "$LINT" == "1" ]]; then
+		# Lint with exit CI
```
Comment (sfan5) [inline util/travis/script.sh]: does this perform linting **instead** of compiling?

Comment (nerzhul) [inline util/travis/script.sh]: exact, it's another build, it doesn't need to be done on each build. When linting will be mandatory on some/all files it permits to know directly the problem is codestyle

File: util/travis/before_install.sh (hunk 4 above)
Comment (nerzhul) [inline util/travis/before_install.sh]: exact, it was needed before but not now

File: util/travis/script.sh (hunk 7 above)
Comment (sfan5) [inline util/travis/script.sh]: then maybe set `CC=none` in the config to make it clear that it doesn't actually compile anything

**Comment by nerzhul** [comment]
@sfan5 is this okay for you ?

**Comment by nerzhul** [comment]
I just added a code style fix to trigger properly lint

**Comment by nerzhul** [comment]
Hmmm seems the find doesn't return the good files to check: https://travis-ci.org/minetest/minetest/jobs/213212543 

i will fix it tomorrow

**Comment by nerzhul** [comment]
Okay it works now properly, just fixing one parameter and we can merge

https://travis-ci.org/minetest/minetest/jobs/213305789

**Comment by nerzhul** [comment]
I did a last cleanup.

See https://travis-ci.org/minetest/minetest/jobs/213431308

**Comment by nerzhul** [comment]
I think this should be merged, and when we get more time we should verify the full rules correctness and add a whitelist of non compliant existing files and remove the lazy move for others

File: util/travis/script.sh [hunk 8]
```diff
@@ -3,21 +3,55 @@
...
+	else
+		# Check everything for branch pushes
+		files_to_lint="$(find src/ -name '*.cpp' -or -name '*.h' | egrep -v '^src/(gmp|lua|jsoncpp)/')"
+	fi
+
+	for f in ${files_to_lint}; do
+		d=$(diff -u "$f" <(${CLANG_FORMAT} "$f") || true)
+		if ! [ -z "$d" ]; then
+			printf "The file %s is not compliant with the coding style:\n%s\n" "$f" "$d"
+			fail=0
```
Comment (nerzhul) [inline util/travis/script.sh]: please note we are in lazy mode at this moment

File: util/travis/common.sh [hunk 9]
```diff
@@ -1,8 +1,9 @@
 #!/bin/bash -e
 
 # Relative to git-repository root:
-TRIGGER_COMPILE_PATHS="src/|CMakeLists.txt|cmake/Modules/|util/travis/|util/buildbot/"
+TRIGGER_COMPILE_PATHS="src/[^.].*[.](c|cpp|h)|CMakeLists.txt|cmake/Modules/|util/travis/|util/buildbot/"
```
Comment (sfan5) [inline util/travis/common.sh]: why the first `[^.]`?
also imo this is easier to understand as `.*\.(c|cpp|h)`

File: util/travis/script.sh [hunk 10]
```diff
@@ -3,21 +3,55 @@
...
+	CLANG_FORMAT=clang-format-3.9
+	if [ "$TRAVIS_EVENT_TYPE" = "pull_request" ]; then
+		# Get list of every file modified in this pull request
+		files_to_lint="$(git diff --name-only --diff-filter=ACMRTUXB $TRAVIS_COMMIT_RANGE | grep '^src/[^.]*[.]\(cpp\|h\)$' | egrep -v '^src/(gmp|lua|jsoncpp)/' || true)"
+	else
+		# Check everything for branch pushes
+		files_to_lint="$(find src/ -name '*.cpp' -or -name '*.h' | egrep -v '^src/(gmp|lua|jsoncpp)/')"
+	fi
+
+	for f in ${files_to_lint}; do
```
Comment (sfan5) [inline util/travis/script.sh]: define
```bash
    local fail=0
```
above here

File: util/travis/script.sh (hunk 8 above)
Comment (sfan5) [inline util/travis/script.sh]: this should be in a comment there not to confuse people

**Pull request merged**